streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java

```diff
@@ -185,7 +181,7 @@ StreamTask createActiveTaskFromStandby(final StandbyTask standbyTask,
         final LogContext logContext = getLogContext(standbyTask.id);
 
         standbyTask.closeCleanAndRecycleState();
-        stateManager.transitionTaskType(TaskType.ACTIVE, logContext);
+        stateManager.prepareNewTaskType(TaskType.ACTIVE, logContext);
```

Comment:
This is another tricky part: Since the restore thread check's the state-manager's type to decide whether to check for completion, for those recycled tasks we should not change their type immediately before they are asynchronously de-registered (and later re-registered) with the restore thread.